app/views/shared/_error_summary.html.erb

```diff
@@ -0,0 +1,28 @@
+<% class_name = class_name.presence || object.class.name.demodulize.underscore.humanize.downcase %>
+
+<% if object.errors.present? %>
+  <div class="gem-c-error-summary govuk-error-summary" data-module="govuk-error-summary" aria-labelledby="error-summary-title" role="alert" id="error-summary" data-govuk-error-summary-module-started="true">
+    <h2 class="govuk-error-summary__title" id="error-summary-title">
+      To save the <%= class_name %> please fix the following issues:
+    </h2>
+
+    <div class="govuk-error-summary__body">
+      <ul class="govuk-list govuk-error-summary__list">
+        <% object.errors.errors.map do |error| %>
+          <% analytics_action = "#{object.class.name.demodulize.underscore.dasherize}-error" %>
+          <% error_message = error.full_message %>
+
+          <%= tag.li(
+            tag.a(
+              error_message,
+              href: "##{object.class.to_s.underscore}_#{error.attribute.to_s}",
+              class: "govuk-link"
+            ),
+            data: track_analytics_data("form-error", analytics_action, error_message),
+            class: "gem-c-error-summary__list-item"
+          ) %>
+        <% end %>
+      </ul>
+    </div>
+  </div>
+<% end %>
```

Comment:
Re: test coverage

It's a shame we can't use [RSpec View specs](https://relishapp.com/rspec/rspec-rails/v/5-1/docs/view-specs/view-spec) here to test this partial. Here's a real-world [example from a previous project](https://github.com/ministryofjustice/offender-management-allocation-manager/blob/d861ca773c8208168b5176065804bea3dac22541/spec/views/shared/case_type_badge.html.erb_spec.rb).

I tried looking to see if MiniTest has anything similar, but as far as I can tell the general approach is to [test views within controller tests](https://guides.rubyonrails.org/testing.html#testing-views) 😕

Reply:
I really strongly feel that we should use the view components gem. Here's a blog github wrote about them https://github.blog/2020-12-15-encapsulating-ruby-on-rails-views/

They're soooo fast compared to controller specs and far easier to use. They're amazing for a bunch if reasons, but in my experience they've led to far far better test coverage due to ease of use and speed.

It's so easy to unit test conditionals in views (which we have tonnes of)

it also leads to not having test files which are 500+ lines long.

They're amazing. They're the best thing i've encountered for improving a codebase (with a frontend) since becoming a dev.

Reply:
@davidgisbey sorry I think I was typing as you were replying. 

Do you have ideas of how an adoption of ViewComponent would work? and/or how'd they compare to testing with ActionView::TestCase?

Seems like it'd be good to have a discussion early into this work where lots of views are changing as to whether we want to carry on as we are, adopt some degree of view testing or go all the way to view components.

Reply:
Yea i think that it makes sense to make any changes now as we're essentially going to eventually rebuild every view in Whitehall in the GOV.UK Design System.

They're not wildly dissimilar to using a view helper and  `ActionView::TestCase`, but imo it's a more comprehensive solution.  It's extremely cheap to test them. Each component encapsulates the required functionality which makes it really easy to use and significantly reduces cognitive complexity.  I've found it's particularly good at testing views with many conditionals

Here's a good example of why they're so great imo from one i previously built.

https://github.com/DFE-Digital/apply-for-teacher-training/blob/main/app/components/candidate_interface/apply_again_call_to_action_component.html.erb

https://github.com/DFE-Digital/apply-for-teacher-training/blob/main/app/components/candidate_interface/apply_again_call_to_action_component.rb

https://github.com/DFE-Digital/apply-for-teacher-training/blob/main/spec/components/candidate_interface/apply_again_call_to_action_component_spec.rb

then it's called like this https://github.com/DFE-Digital/apply-for-teacher-training/blob/fe2f15fd513a531dbc5899505eb95c8716284258/app/components/candidate_interface/application_dashboard_component.html.erb#L11

There's so much great stuff about them. Would love to chat about it at some point.

Whichever way we go though i think we should probs decide quite quickly. Either solution will be far better than using view_tests in the controller though in terms of ease of use and performance.

Reply:
Great - do you want to have a think about putting in a chat in with Ollie and me to explore some ideas?

From my perspective, I like view components but worry about creating fresh inconsistencies and surprises, so we'd have to trade that off. Similarly I don't like it if we feel we need a cheap way to test views and don't have it.

Reply:
Sounds good i'll pop something in for tomorrow.

Reply:
We've discussed this. I'm going to spike it out and we're going to do a write up with recommendations